src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentSynchronization/RazorDidChangeTextDocumentEndpoint.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.EndpointContracts;
+using Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.DocumentSynchronization;
+
+internal class RazorDidChangeTextDocumentEndpoint : IVSDidChangeTextDocumentEndpoint
+{
+    public bool MutatesSolutionState => true;
+
+    private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+    private readonly RazorProjectService _projectService;
+
+    public RazorDidChangeTextDocumentEndpoint(
+        ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+        RazorProjectService razorProjectService)
+    {
+        _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+        _projectService = razorProjectService;
+    }
+
+    public RegistrationExtensionResult GetRegistration(VSInternalClientCapabilities clientCapabilities)
+    {
+        const string AssociatedServerCapability = "textDocumentSync";
+        var registrationOptions = new TextDocumentSyncOptions()
+        {
+            Change = TextDocumentSyncKind.Incremental,
+            OpenClose = true,
+            Save = new SaveOptions()
+            {
+                IncludeText = true,
+            },
+            WillSave = false,
+            WillSaveWaitUntil = false,
+        };
+
+        var result = new RegistrationExtensionResult(AssociatedServerCapability, registrationOptions);
+
+        return result;
+    }
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(DidChangeTextDocumentParamsBridge request)
+    {
+        return request.TextDocument;
+    }
+
+    public async Task HandleNotificationAsync(DidChangeTextDocumentParamsBridge request, RazorRequestContext context, CancellationToken cancellationToken)
+    {
+        var documentContext = context.GetRequiredDocumentContext();
+
+        var sourceText = await documentContext.GetSourceTextAsync(cancellationToken);
+        sourceText = ApplyContentChanges(request.ContentChanges, sourceText, context.Logger);
+
+        await _projectSnapshotManagerDispatcher.RunOnDispatcherThreadAsync(
+            () => _projectService.UpdateDocument(documentContext.FilePath, sourceText, request.TextDocument.Version),
+            CancellationToken.None).ConfigureAwait(false);
```

Comment:
The existing code [does not pass a cancellationToken](https://github.com/dotnet/razor-tooling/blob/a18b6d3577c98e99e4d38125eac20433009b5bd5/src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDocumentSynchronizationEndpoint.cs#L70). I assumed this was done on purpose to ensure that as long as we got this far the document would update.

Reply:
I think in general you chain the cancellation unless cancelling the operation in question has ways to leave state in an unrecoverable/irreversible shape. In this case the update operation is a simple array update after state has been evaluated. So no reason not to pass it in.

Reply:
Also curious if the document version is really useful outside the testing scenarios. The document operation all have inbuilt no-op check optimizations, which seems to be done for the purpose of tracking the document cache version. But is the version itself useful?

Reply:
The document version is just an incrementing integer, not useful for anything but comparisons. With CLaSP and single server, I would think/hope/desire that we can get rid of document versions altogether.

As for cancellation, I can see why there would be a worry here - if the server sends us a didChange and we don't apply it, then we're out of sync with the client. I'm just not sure thats a realistic thing that would happen, because who would be cancelling us, except the client, and if the client tries to cancel us and we _don't_ pass in the token, then we're out of sync anyway!